clippy_lints/src/loops/unused_enumerate_index.rs

```diff
@@ -0,0 +1,76 @@
+use super::UNUSED_ENUMERATE_INDEX;
+use clippy_utils::diagnostics::{multispan_sugg, span_lint_and_then};
+use clippy_utils::source::snippet;
+use clippy_utils::sugg;
+use clippy_utils::visitors::is_local_used;
+use rustc_hir::def::DefKind;
+use rustc_hir::{Expr, ExprKind, Pat, PatKind};
+use rustc_lint::LateContext;
+use rustc_middle::ty;
+
+/// Checks for the `UNUSED_ENUMERATE_INDEX` lint.
+pub(super) fn check<'tcx>(cx: &LateContext<'tcx>, pat: &'tcx Pat<'_>, arg: &'tcx Expr<'_>, body: &'tcx Expr<'_>) {
+    let PatKind::Tuple(tuple, _) = pat.kind else {
+        return;
+    };
+
+    let ExprKind::MethodCall(method, self_arg, [], _) = arg.kind else {
+        return;
+    };
+
+    if method.ident.as_str() != "enumerate" {
+        return;
+    }
+
+    let ty = cx.typeck_results().expr_ty(arg);
+
+    if !pat_is_wild(cx, &tuple[0].kind, body) {
+        return;
+    }
+
+    let name = match *ty.kind() {
+        ty::Adt(base, _substs) => cx.tcx.def_path_str(base.did()),
+        _ => return,
+    };
+
+    if name != "std::iter::Enumerate" && name != "core::iter::Enumerate" {
+        return;
+    }
+
+    let Some((DefKind::AssocFn, call_id)) = cx.typeck_results().type_dependent_def(arg.hir_id) else {
+        return;
+    };
+
+    let call_name = cx.tcx.def_path_str(call_id);
+
+    if call_name != "std::iter::Iterator::enumerate" && call_name != "core::iter::Iterator::enumerate" {
```

Comment:
Also another corner case: what if the `impl Iterator for I` overrides the `fn enumerate(...)`? How should such cases be handled? rustdoc doesn't seem to point out such known overrides (or at least I don't know where to look), so are there any other overrides that should be added here?

Reply:
AFAIK, there's nothing that we can really do to know if a method has been overriden that wouldn't produce false negatives. Also, it seems like [it isn't a common pattern to override `enumerate`](https://github.com/search?q=lang%3Ars+fn+enumerate%28self%29+-%3E+Enumerate%3CSelf%3E&type=code). So overriden `enumerate` isn't really a problem nor has solution.

Reply:
There's no way to override any of the adapter methods without unsafe code. The return types don't have any exposed constructors.